File: src/crypto.js
// Web Crypto. The entropy, the digests and the two key algorithms are the
// host's; a key's material lives here, as hex, and goes back down for every
// operation.
(function () {
  'use strict';

  const INTEGER_VIEWS = [
    Int8Array,
    Uint8Array,
    Uint8ClampedArray,
    Int16Array,
    Uint16Array,
    Int32Array,
    Uint32Array,
    BigInt64Array,
    BigUint64Array,
  ];

  function randomBytes(count) {
    const hex = __ow_native_random_hex(count);
    const bytes = new Uint8Array(count);

    for (let i = 0; i < count; i++) {
      bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }

    return bytes;
  }

  const DIGESTS = ['SHA-1', 'SHA-256', 'SHA-384', 'SHA-512'];

  function bytesOf(source) {
    if (source instanceof ArrayBuffer) {
      return new Uint8Array(source);
    }

    if (ArrayBuffer.isView(source)) {
      return new Uint8Array(source.buffer, source.byteOffset, source.byteLength);
    }

    throw new TypeError('the data to digest has to be a buffer or a view of one');
  }

  function toHex(bytes) {
    let hex = '';

    for (const byte of bytes) {
      hex += byte.toString(16).padStart(2, '0');
    }

    return hex;
  }

  function fromHex(hex) {
    const bytes = new Uint8Array(hex.length / 2);

    for (let i = 0; i < bytes.length; i++) {
      bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }

    return bytes;
  }

  // Key material, kept off the object the guest holds.
  const material = new WeakMap();

  function algorithmName(algorithm) {
    return String(
      algorithm !== null && typeof algorithm === 'object' ? algorithm.name : algorithm
    ).toUpperCase();
  }

  class CryptoKey {
    constructor(type, algorithm, extractable, usages, raw) {
      this.type = type;
      this.algorithm = algorithm;
      this.extractable = extractable;
      this.usages = usages;

      material.set(this, raw);
    }
  }

  function rawOf(key) {
    const raw = material.get(key);

    if (raw === undefined) {
      throw new DOMException('that is not a key this runtime made', 'InvalidAccessError');
    }

    return raw;
  }

  function hashOf(algorithm, key) {
    const hash = algorithm !== null && typeof algorithm === 'object' ? algorithm.hash : undefined;

    return String(hash === undefined ? key.algorithm.hash : hash).toUpperCase();
  }

  class SubtleCrypto {
    async importKey(format, data, algorithm, extractable, usages) {
      if (String(format) !== 'raw') {
        throw new DOMException(format + ' keys are not imported here', 'NotSupportedError');
      }

      const name = algorithmName(algorithm);

      if (name !== 'HMAC' && name !== 'AES-GCM') {
        throw new DOMException(name + ' is not a key algorithm here', 'NotSupportedError');
      }

      const shape =
        name === 'HMAC'
          ? { name: name, hash: String(algorithm.hash).toUpperCase() }
          : { name: name, length: bytesOf(data).byteLength * 8 };

      return new CryptoKey('secret', shape, Boolean(extractable), [...usages], toHex(bytesOf(data)));
    }

    async generateKey(algorithm, extractable, usages) {
      const name = algorithmName(algorithm);

      if (name !== 'AES-GCM') {
        throw new DOMException(name + ' pairs are not generated here', 'NotSupportedError');
      }

      const length = Number(algorithm.length) || 256;

      if (length !== 256) {
        throw new DOMException('AES-GCM here takes a 256 bit key', 'NotSupportedError');
      }

      return new CryptoKey(
        'secret',
        { name: name, length: length },
        Boolean(extractable),
        [...usages],
        toHex(randomBytes(length / 8))
      );
    }

    async exportKey(format, key) {
      if (String(format) !== 'raw') {
        throw new DOMException(format + ' keys are not exported here', 'NotSupportedError');
      }

      if (!key.extractable) {
        throw new DOMException('that key was imported as not extractable', 'InvalidAccessError');
      }

      return fromHex(rawOf(key)).buffer;
    }

    async sign(algorithm, key, data) {
      if (algorithmName(algorithm) !== 'HMAC') {
        throw new DOMException('only HMAC signs here', 'NotSupportedError');
      }

      const tag = __ow_native_hmac(hashOf(algorithm, key), rawOf(key), toHex(bytesOf(data)));

      if (tag === null) {
        throw new DOMException('that is not a hash this runtime has', 'NotSupportedError');
      }

      return fromHex(tag).buffer;
    }

    async verify(algorithm, key, signature, data) {
      if (algorithmName(algorithm) !== 'HMAC') {
        throw new DOMException('only HMAC verifies here', 'NotSupportedError');
      }

      return __ow_native_hmac(
        hashOf(algorithm, key),
        rawOf(key),
        toHex(bytesOf(data)),
        toHex(bytesOf(signature))
      ) === true;
    }

    async encrypt(algorithm, key, data) {
      return aes('encrypt', algorithm, key, data);
    }

    async decrypt(algorithm, key, data) {
      return aes('decrypt', algorithm, key, data);
    }

    async digest(algorithm, data) {
      const name = String(
        algorithm !== null && typeof algorithm === 'object' ? algorithm.name : algorithm
      ).toUpperCase();

      if (!DIGESTS.includes(name)) {
        throw new DOMException(name + ' is not a digest this runtime has', 'NotSupportedError');
      }

      return fromHex(__ow_native_digest(name, toHex(bytesOf(data)))).buffer;
    }
  }

  function aes(op, algorithm, key, data) {
    if (algorithmName(algorithm) !== 'AES-GCM') {
      throw new DOMException('only AES-GCM is a cipher here', 'NotSupportedError');
    }

    const out = __ow_native_aes_gcm(
      op,
      rawOf(key),
      toHex(bytesOf(algorithm.iv)),
      toHex(bytesOf(data))
    );

    if (out === null) {
      throw new DOMException('AES-GCM here takes a 256 bit key and a 12 byte iv', 'DataError');
    }

    if (out === false) {
      throw new DOMException('the tag does not hold', 'OperationError');
    }

    return fromHex(out).buffer;
  }

  class Crypto {
    getRandomValues(view) {
      if (!INTEGER_VIEWS.some((kind) => view instanceof kind)) {
        throw new DOMException(
          'getRandomValues expects an integer typed array',
          'TypeMismatchError'
        );
      }

      if (view.byteLength > 65536) {
        throw new DOMException(
          'getRandomValues accepts at most 65536 bytes',
          'QuotaExceededError'
        );
      }

      const target = new Uint8Array(view.buffer, view.byteOffset, view.byteLength);

      target.set(randomBytes(view.byteLength));

      return view;
    }

    randomUUID() {
      const bytes = randomBytes(16);

      bytes[6] = (bytes[6] & 0x0f) | 0x40;
      bytes[8] = (bytes[8] & 0x3f) | 0x80;

      const hex = Array.from(bytes, (byte) =>
        byte.toString(16).padStart(2, '0')
      ).join('');

      return [
        hex.slice(0, 8),
        hex.slice(8, 12),
        hex.slice(12, 16),
        hex.slice(16, 20),
        hex.slice(20),
      ].join('-');
    }

    get subtle() {
      return subtle;
    }
  }

  const subtle = new SubtleCrypto();

  globalThis.Crypto = Crypto;
  globalThis.CryptoKey = CryptoKey;
  globalThis.SubtleCrypto = SubtleCrypto;
  globalThis.crypto = new Crypto();
})();

File: src/crypto.rs
//! The entropy source behind the guest's `crypto`, and the digests behind its
//! `subtle`.

use nova_vm::ecmascript::Agent;
use nova_vm::ecmascript::ArgumentsList;
use nova_vm::ecmascript::JsResult;
use nova_vm::ecmascript::String as JsString;
use nova_vm::ecmascript::Value;
use nova_vm::engine::Bindable;
use nova_vm::engine::GcScope;
use nova_vm::engine::Scopable;

use aes_gcm::Aes256Gcm;
use aes_gcm::KeyInit;
use aes_gcm::aead::Aead;
use aes_gcm::aead::Payload;

use hmac::Hmac;
use hmac::Mac;

use sha1::Sha1;
use sha2::Digest;
use sha2::Sha256;
use sha2::Sha384;
use sha2::Sha512;

/// The largest request `crypto.getRandomValues` accepts, so a guest reaching
/// past the JS layer cannot ask for an unbounded allocation.
const MAX_BYTES: i32 = 65_536;

/// `__ow_native_random_hex(count)`: hex-encoded random bytes.
pub fn native_random_hex<'gc>(
    agent: &mut Agent,
    _this: Value,
    args: ArgumentsList,
    mut gc: GcScope<'gc, '_>,
) -> JsResult<'gc, Value<'gc>> {
    let count = args.get(0).to_int32(agent, gc.reborrow()).unbind()?;
    let count = count.clamp(0, MAX_BYTES) as usize;

    let mut bytes = vec![0u8; count];

    getrandom::fill(&mut bytes).expect("the system random source is always available");

    let mut hex = String::with_capacity(count * 2);

    for byte in bytes {
        hex.push_str(&format!("{byte:02x}"));
    }

    Ok(JsString::from_string(agent, hex, gc.nogc()).unbind().into())
}

/// `__ow_native_digest(algorithm, hex)`: the digest of those bytes, in hex, or
/// `null` for an algorithm this runtime does not have. Hex both ways because
/// nova_vm gives an embedder no way to build or read a typed array.
pub fn native_digest<'gc>(
    agent: &mut Agent,
    _this: Value,
    args: ArgumentsList,
    mut gc: GcScope<'gc, '_>,
) -> JsResult<'gc, Value<'gc>> {
    let input = args.get(1).scope(agent, gc.nogc());

    let algorithm = args.get(0).to_string(agent, gc.reborrow()).unbind()?;
    let algorithm = algorithm.to_string_lossy(agent).into_owned();

    let input = input.get(agent).to_string(agent, gc.reborrow()).unbind()?;
    let input = input.to_string_lossy(agent).into_owned();

    let Some(bytes) = from_hex(&input) else {
        return Ok(Value::Null);
    };

    let digest = match algorithm.as_str() {
        "SHA-1" => to_hex(&Sha1::digest(&bytes)),
        "SHA-256" => to_hex(&Sha256::digest(&bytes)),
        "SHA-384" => to_hex(&Sha384::digest(&bytes)),
        "SHA-512" => to_hex(&Sha512::digest(&bytes)),
        _ => return Ok(Value::Null),
    };

    Ok(JsString::from_string(agent, digest, gc.nogc())
        .unbind()
        .into())
}

fn from_hex(text: &str) -> Option<Vec<u8>> {
    if !text.len().is_multiple_of(2) {
        return None;
    }

    (0..text.len())
        .step_by(2)
        .map(|at| u8::from_str_radix(&text[at..at + 2], 16).ok())
        .collect()
}

fn to_hex(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

/// `__ow_native_hmac(hash, keyHex, dataHex, tagHex)`: the tag, in hex, or
/// `"true"`/`"false"` when a tag is given to check against. Verification runs
/// here so the comparison is the constant-time one the crate does.
pub fn native_hmac<'gc>(
    agent: &mut Agent,
    _this: Value,
    args: ArgumentsList,
    mut gc: GcScope<'gc, '_>,
) -> JsResult<'gc, Value<'gc>> {
    let key = args.get(1).scope(agent, gc.nogc());
    let data = args.get(2).scope(agent, gc.nogc());
    let tag = args.get(3).scope(agent, gc.nogc());

    let hash = args.get(0).to_string(agent, gc.reborrow()).unbind()?;
    let hash = hash.to_string_lossy(agent).into_owned();

    let key = key.get(agent).to_string(agent, gc.reborrow()).unbind()?;
    let key = key.to_string_lossy(agent).into_owned();

    let data = data.get(agent).to_string(agent, gc.reborrow()).unbind()?;
    let data = data.to_string_lossy(agent).into_owned();

    let tag = tag.get(agent);
    let tag = if tag.is_undefined() {
        None
    } else {
        let tag = tag.to_string(agent, gc.reborrow()).unbind()?;

        Some(tag.to_string_lossy(agent).into_owned())
    };

    let (Some(key), Some(data)) = (from_hex(&key), from_hex(&data)) else {
        return Ok(Value::Null);
    };

    macro_rules! mac {
        ($hash:ty) => {{
            let mut mac = <Hmac<$hash> as KeyInit>::new_from_slice(&key)
                .expect("HMAC takes a key of any length");

            mac.update(&data);
            mac
        }};
    }

    if let Some(tag) = tag {
        let Some(tag) = from_hex(&tag) else {
            return Ok(Value::Boolean(false));
        };

        let held = match hash.as_str() {
            "SHA-256" => mac!(Sha256).verify_slice(&tag).is_ok(),
            "SHA-384" => mac!(Sha384).verify_slice(&tag).is_ok(),
            "SHA-512" => mac!(Sha512).verify_slice(&tag).is_ok(),
            "SHA-1" => mac!(Sha1).verify_slice(&tag).is_ok(),
            _ => return Ok(Value::Null),
        };

        return Ok(Value::Boolean(held));
    }

    let answer = match hash.as_str() {
        "SHA-256" => to_hex(&mac!(Sha256).finalize().into_bytes()),
        "SHA-384" => to_hex(&mac!(Sha384).finalize().into_bytes()),
        "SHA-512" => to_hex(&mac!(Sha512).finalize().into_bytes()),
        "SHA-1" => to_hex(&mac!(Sha1).finalize().into_bytes()),
        _ => return Ok(Value::Null),
    };

    Ok(JsString::from_string(agent, answer, gc.nogc())
        .unbind()
        .into())
}

/// `__ow_native_aes_gcm(op, keyHex, ivHex, dataHex)`: the ciphertext with its
/// tag appended, or the plaintext, in hex. `null` when the key or the nonce is
/// the wrong size, and `false` when the tag does not hold.
pub fn native_aes_gcm<'gc>(
    agent: &mut Agent,
    _this: Value,
    args: ArgumentsList,
    mut gc: GcScope<'gc, '_>,
) -> JsResult<'gc, Value<'gc>> {
    let key = args.get(1).scope(agent, gc.nogc());
    let iv = args.get(2).scope(agent, gc.nogc());
    let data = args.get(3).scope(agent, gc.nogc());

    let op = args.get(0).to_string(agent, gc.reborrow()).unbind()?;
    let op = op.to_string_lossy(agent).into_owned();

    let key = key.get(agent).to_string(agent, gc.reborrow()).unbind()?;
    let key = key.to_string_lossy(agent).into_owned();

    let iv = iv.get(agent).to_string(agent, gc.reborrow()).unbind()?;
    let iv = iv.to_string_lossy(agent).into_owned();

    let data = data.get(agent).to_string(agent, gc.reborrow()).unbind()?;
    let data = data.to_string_lossy(agent).into_owned();

    let (Some(key), Some(iv), Some(data)) = (from_hex(&key), from_hex(&iv), from_hex(&data)) else {
        return Ok(Value::Null);
    };

    // A slice no longer converts on its own; the sizes are the ones AES-256-GCM
    // takes, so the conversion is the check.
    let (Ok(key), Ok(iv)) = (
        <&[u8; 32]>::try_from(key.as_slice()),
        <&[u8; 12]>::try_from(iv.as_slice()),
    ) else {
        return Ok(Value::Null);
    };

    let cipher = Aes256Gcm::new(key.into());
    let nonce = iv.into();
    let payload = Payload {
        msg: &data,
        aad: &[],
    };

    let out = match op.as_str() {
        "encrypt" => cipher.encrypt(nonce, payload).ok(),
        "decrypt" => cipher.decrypt(nonce, payload).ok(),
        _ => return Ok(Value::Null),
    };

    let Some(out) = out else {
        return Ok(Value::Boolean(false));
    };

    Ok(JsString::from_string(agent, to_hex(&out), gc.nogc())
        .unbind()
        .into())
}
